src/decision/location_selection.jl

```diff
@@ -0,0 +1,298 @@
+    end
+
+    return rank_frequencies
+end
+function ranks_to_frequencies(
+    ranks::NamedDimsArray{D,T,3,A};
+    n_ranks::Int64=length(ranks.sites),
+    agg_func=nothing,
+) where {D,T,A}
+    if !isnothing(agg_func)
+        return agg_func(ranks_to_frequencies(ranks, n_ranks))
+    end
+
+    return ranks_to_frequencies(ranks, n_ranks)
+end
+function ranks_to_frequencies(
+    ranks::NamedDimsArray{D,T,2,A};
+    n_ranks::Int64=length(ranks.sites),
+    agg_func=nothing,
+) where {D,T,A}
+    if !isnothing(agg_func)
+        return agg_func(ranks_to_frequencies(ranks, n_ranks))
+    end
+
+    return ranks_to_frequencies(ranks, n_ranks)
+end
+
+"""
+    location_selection_frequencies(ranks::NamedDimsArray; n_loc_int::Int64=5)
+    location_selection_frequencies(iv_log::NamedDimsArray{D,T,4,A}; dims::Union{Symbol,Vector{Symbol}}=:coral_id) where {D,T,A}
+
+Determines the count of times each location was selected for a specific intervention over a
+set of scenarios.
+
+# Arguments
+- `ranks` : Rankings of locations `rank_locations()`
+- `n_loc_int` : number of locations intervened at, for each decision point
+- `iv_log` : Intervention logs
+- `dims` : dimensions to sum selection frequencies over
+
+# Returns
+Number of times each location was selected for an intervention.
+"""
+function location_selection_frequencies(
+    ranks::NamedDimsArray;
+    n_loc_int::Int64=5,
+)
+    ranks_frequencies = ranks_to_frequencies(ranks; n_ranks=n_loc_int)
+    loc_count = sum(ranks_frequencies[ranks=1:n_loc_int], dims=2)[ranks=1]
+
+    return loc_count
+end
+function location_selection_frequencies(
+    iv_log::NamedDimsArray{D,T,4,A};
+    dims::Union{Symbol,Vector{Symbol}}=:coral_id,
+) where {D,T,A}
+    loc_count = dropdims(
+        sum(dropdims(sum(iv_log; dims=dims); dims=dims) .> 0; dims=:scenarios);
+        dims=:scenarios,
+    )
+    return loc_count
+end
+
+"""Drop single dimensions."""
+_drop_single(x::AbstractMatrix) = dropdims(x, dims=(findall(size(x) .== 1)...,))
+
+"""
+    selection_score(ranks::NamedDimsArray{D,T,3,A}; dims::Vector{Symbol}=[:scenarios, :timesteps]) where {D,T,A}
+    selection_score(ranks::NamedDimsArray{D,T,2,A}) where {D,T,A}
+    selection_score(ranks::NamedDimsArray, dims::Vector{Symbol})
+
+Calculates score ∈ [0, 1], where 1 is the highest score possible, indicative of the relative
+desirability of each location.
+
+The score reflects the location ranking and frequency of attaining a high rank.
+
+# Arguments
+- `ranks` : Rankings of locations from `rank_locations()`
+- `dims` : Dimensions to sum over
+
+# Returns
+Selection score
+"""
+function selection_score(
+    ranks::NamedDimsArray{D,T,3,A};
+    dims::Vector{Symbol}=[:scenarios, :timesteps],
+) where {D,T,A}
+    return _drop_single(selection_score(ranks, dims))
+end
+function selection_score(
+    ranks::NamedDimsArray{D,T,2,A};
+    dims::Vector{Symbol}=[:scenarios],
+) where {D,T,A}
+    return _drop_single(selection_score(ranks, dims))
```

Comment:
I tried to run the code from the cookbook and I got an error in this line when running `sel_score = selection_score(ranks[intervention=1])`:

```julia
ERROR: MethodError: no method matching _drop_single(::NamedDims.NamedDimsArray{(:sites,), Float64, 1, AxisKeys.KeyedArray{Float64, 1, Vector{Float64}, Base.RefValue{UnitRange{Int64}}}})

Closest candidates are:
  _drop_single(::AbstractMatrix)
   @ ADRIA C:\Users\pribeiro\AIMS\Code\ADRIA.jl\src\decision\location_selection.jl:260

Stacktrace:
 [1] selection_score(ranks::NamedDims.NamedDimsArray{(:sites, :scenarios), Float64, 2, AxisKeys.KeyedArray{Float64, 2, Matrix{Float64}, Tuple{UnitRange{Int64}, UnitRange{Int64}}}}; dims::Vector{Symbol})
   @ ADRIA C:\Users\pribeiro\AIMS\Code\ADRIA.jl\src\decision\location_selection.jl:289
 [2] selection_score(ranks::NamedDims.NamedDimsArray{(:sites, :scenarios), Float64, 2, AxisKeys.KeyedArray{Float64, 2, Matrix{Float64}, Tuple{UnitRange{Int64}, UnitRange{Int64}}}})
   @ ADRIA C:\Users\pribeiro\AIMS\Code\ADRIA.jl\src\decision\location_selection.jl:285
 [3] top-level scope
   @ c:\Users\pribeiro\AIMS\Code\ADRIA.jl\sandbox\cookbook\loc_agg_funcs.jl:36
```

Which leads me to: it would be better if the returning types were explicit. (In this case it would make debugging easier).

Reply:
I also got the same error when running `rank_frequencies_seed = rank_locations(
    dom, scens, sum_cover, area_to_seed, selection_score, 1
)` also from the cookbook.

Reply:
Could you try pulling again Pedro? I pushed some more changes after I asked you to review

Reply:
There was no change since the version I had when I tested this - which was about an hour ago.